sample-code/src/tutorials/agent-workflow-openai-langchain.ts

```diff
@@ -0,0 +1,135 @@
+const tools = [getWeatherTool, getRestaurantsTool];
+const toolNode = new ToolNode(tools);
+
+// Create a model and give it access to the tools
+const model = new AzureOpenAiChatClient({
+  modelName: 'gpt-4o',
+  temperature: 0.7,
+}).bindTools(tools);
+
+async function shouldContinueAgent({ messages }: typeof MessagesAnnotation.State) {
+  const lastMessage = messages[messages.length - 1] as AIMessage;
+  
+  // If there are tool calls, go to tools
+  if (lastMessage.tool_calls?.length) {
+    return 'tools';
+  }
+  // Check if agent's message is a farewell (indicating user was satisfied)
+  const result = await model.invoke([
+    new SystemMessage(`You are a classifier. Respond with exactly "FAREWELL" if this is a farewell/goodbye message wishing someone happy travels. Respond with exactly "CONTINUE" if the conversation should continue.`),
+    new HumanMessage(`Assistant message: "${lastMessage.content}"`)
+  ]);
+  
+  return result.content === 'FAREWELL' ? END : 'askHuman';
+}
+
+async function askHuman({ messages }: typeof MessagesAnnotation.State) {
+  // This is where the actual interrupt happens
+  const humanResponse: string = interrupt(`Do you want to adjust the itinerary?`);
+  return { messages: [new HumanMessage(humanResponse)] };
+}
+
+async function callModel(state: typeof MessagesAnnotation.State) {
+  const response = await model.invoke(state.messages);
+  return { messages: [response] };
+}
+
+  const workflow = new StateGraph(MessagesAnnotation)
+  .addNode("agent", callModel)
+  .addNode("tools", toolNode)
+  .addNode("askHuman", askHuman)
+  .addConditionalEdges("agent", shouldContinueAgent, ["tools", "askHuman", END])
+  .addEdge("tools", "agent")
+  .addEdge("askHuman", "agent")
+  .addEdge(START, "agent");
+
+const memory = new MemorySaver();
+const app = workflow.compile({ checkpointer: memory });
+
+  const config = { configurable: { thread_id: 'conv-1' } };
+
+export async function runTravelAssistant() {
+    // Initial system prompt and user message
+    let messages = [
+      new SystemMessage(
+        `You are a helpful travel assistant.  
+        You will generate a 3-item itinerary based on a provided city. You should use weather forecast and restaurant recommendations when creating the itinerary.
+        After presenting the itinerary, ask the user if they are satisfied with it or if they want to make changes.`
+      ),
+      new HumanMessage(
+        "I'm traveling to Paris. Can you help me prepare an itinerary?"
+      )
+    ];
+  
+    try {
+      let response = await app.invoke({ messages }, config);
+     
+      console.log('Assistant:', response.messages[response.messages.length - 1].content);
+      console.log("next: ", (await app.getState(config)).next);
+    
+      response = await app.invoke(new Command({ resume: 'Can you suggest something more outdoorsy?' }), config);
+
+      console.log('Assistant:', response.messages[response.messages.length - 1].content);
+
```

Comment:
[q] A `next` here as well?

Reply:
Was conflicted about it, whether we need to show the use of next again, since I already explained what this is for before. If you have a strong opinion, I can add it back